Steer/DigitizerWorkflow/src/CTPDigitizerSpec.cxx

```diff
@@ -0,0 +1,132 @@
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/DataProcessorSpec.h"
+#include "Framework//Task.h"
+#include "DetectorsBase/BaseDPLDigitizer.h"
+#include "DataFormatsCTP/Digits.h"
+#include "Steer/HitProcessingManager.h" // for DigitizationContext
+#include "DetectorsCommonDataFormats/DetID.h"
+#include "CTPSimulation/Digitizer.h"
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsFV0/BCData.h"
+
+#include <TStopwatch.h>
+
+using namespace o2::framework;
+namespace o2
+{
+namespace ctp
+{
+class CTPDPLDigitizerTask : public o2::base::BaseDPLDigitizer
+{
+  using GRP = o2::parameters::GRPObject;
+
+ public:
+  CTPDPLDigitizerTask() : o2::base::BaseDPLDigitizer(), mDigitizer() {}
+  ~CTPDPLDigitizerTask() override = default;
+  void initDigitizerTask(framework::InitContext& ic) override
+  {
+    mDigitizer.init();
+  }
+  void run(framework::ProcessingContext& pc)
+  {
+    if (mFinished) {
+      return;
+    }
+    // read collision context from input
+    //auto context = pc.inputs().get<o2::steer::DigitizationContext*>("collisioncontext");
+    //const bool withQED = context->isQEDProvided();
+    //auto& timesview = context->getEventRecords(withQED);
+    // read ctp inputs from input
+    auto ft0inputs = pc.inputs().get<std::vector<o2::ft0::DetTrigInput>>("ft0");
+    auto fv0inputs = pc.inputs().get<std::vector<o2::fv0::DetTrigInput>>("fv0");
+
+    // if there is nothing to do ... return
+    if ((ft0inputs.size() == 0) && (fv0inputs.size() == 0)) {
+      return;
+    }
+    std::map<o2::InteractionRecord, o2::ctp::CTPDigit> finputs;
+    TStopwatch timer;
+    timer.Start();
+    LOG(INFO) << "CALLING CTP DIGITIZATION";
+    //for (const auto& coll : timesview) {
+    //  mDigitizer.setInteractionRecord(coll);
+    //  mDigitizer.process(mDigits);
+    //  mDigitizer.flush(mDigits);
+    //}
+    for (const auto& inp : ft0inputs) {
+      CTPInputDigit finpdigit;
+      finpdigit.mDetector = o2::detectors::DetID::FT0;
+      finpdigit.mInputsMask = inp.mInputs;
+      CTPDigit fctpdigit;
+      fctpdigit.mIntRecord = inp.mIntRecord;
+      fctpdigit.mInputs.push_back(finpdigit);
+      finputs[inp.mIntRecord] = fctpdigit;
+    }
+    for (const auto& inp : fv0inputs) {
+      CTPInputDigit finpdigit;
+      finpdigit.mDetector = o2::detectors::DetID::FV0;
+      finpdigit.mInputsMask = inp.mInputs;
+      if (finputs.count(inp.mIntRecord) == 0) {
+        CTPDigit fctpdigit;
+        fctpdigit.mIntRecord = inp.mIntRecord;
+        fctpdigit.mInputs.push_back(finpdigit);
+        finputs[inp.mIntRecord] = fctpdigit;
+      } else {
+        finputs[inp.mIntRecord].mInputs.push_back(finpdigit);
+      }
+    }
+    for (const auto& inps : finputs) {
+      mDigitizer.setInteractionRecord(inps.first);
+      mDigitizer.process(inps.second, mDigits);
+      mDigitizer.flush(mDigits);
+    }
+    // send out to next stage
+    pc.outputs().snapshot(Output{"CTP", "DIGITS", 0, Lifetime::Timeframe}, mDigits);
```

Comment:
Is your output simply a collection of triggerInputs from detectors? I thought you should apply some logic and generate a decision?

Reply:
Yes, good question. I need to calculate trigger classes . My thoughts are following.
I can do it either here in Digitizer or later in Digit to Raw transformation.
In either case trigger classes are function of trigger inputs depending the CTP configuration.
So we would need to create CTP configuration in GRPC ?
Please, suggestions and examples.
========
PS other:
I can not compile at the moment - it fails on GEANT4_VMC. 
1.) is always branch dev  able to compile ?
2.) how can I get ground state for aliBuild compilation, e.g. if I remove BUILD is it enough ?

Reply:
opps, why so big and bold ?

Reply:
Digit->raw transformation is important for the code validation, full system test etc., but it will not be done in the routine MC productions. So, I think you should calculate the CTP response in the digitization.

> In either case trigger classes are function of trigger inputs depending the CTP configuration.
> So we would need to create CTP configuration in GRPC ?

I guess you mean GRP object? Why not CTP-specifics objects line in Run2? I assume for the anchored MC productions the CCDB data object will be used. For the tests, you can created a dummy one in the `http://ccdb-test.cern.ch:8080` (see https://github.com/AliceO2Group/AliceO2/tree/dev/CCDB if you haven't seen it yet). 
As an example of ccdb configuration creation you can see the ZDC [macro ](https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/ZDC/macro/CreateModuleConfig.C).

>I can not compile at the moment - it fails on GEANT4_VMC.
>1.) is always branch dev able to compile ?

Yes, dev is always compilable (exceptionally, it could be broken but is quickly fixed. Right now there plenty of green CI tests, so there is no problem.) If the installation problem persists, the most efficient is to post the description on https://alice-talk.web.cern.ch/

> 2.) how can I get ground state for aliBuild compilation, e.g. if I remove BUILD is it enough ?

The cleanest will be to remove the sw. Suppressing BUILD should guarantee a clean rebuild, but will leave some orphan dirs. in other subdirs of sw.

Reply:
The database is not implemented yet. Trigger classes calculation is implemented in Digitizer.
Please, let me know if 1.) overall structure is ok 2.) if memory issues (gpl::scan/vectors usage) are ok.